### ISSUE TITLE
[umbrella issue] bash in kubernetes

### ISSUE THREAD

**Issue by cblecker** [kubernetes/kubernetes#78702]
So Apple is deprecating bash as their default shell in their upcoming macOS 10.16, and moving to zsh. We have had many problems supporting multiple bash versions, and I think might be the right cue to move away from bashisms and onto either python or go code.

We had a lengthy discussion in Slack about this, and the rough draft plan is:
- Establish firm ownership of `hack/` and `build/` (it's fuzzy testing/release world right now)
- Write a style guide
- Make python code sustainable
- Find what doesn't conform
- Migrate bash stuff to go/python

In summary, the goals would be:
- POSIX `sh` is okay for short, simple, or bootstrapping scripts
- Anything that's more complicated (function calls, arrays, etc) should be either python or go
- We need dependency management for python

cc: @Katharine @BenTheElder @michelle192837 @thockin @dims @tpepper @stevekuznetsov

**Comment by cblecker** [comment]
Slack discussion: https://kubernetes.slack.com/archives/C09QZ4DQB/p1559666693008200

**Comment by BenTheElder** [comment]
[adding a note that we've also recently had a bunch of issues with bash version compatibility issues on linux, on really commonly installed versions]

For a lot of this tooling, if we do this, we should really consider wholly new tooling. E.G. we have a lot of complexity in the build attempting to "cache" that predates go 1.10, or the rsync stuff which is a bit less necessary now.

**Comment by thockin** [comment]
The rsync stuff is because MacOS performance is terrible for docker volumes.  I would love to drop it, but last I looked we just couldn't.

I don't disagree that bash may be at the end of it's runway, but I question python as an alternative.

**Comment by BenTheElder** [comment]
> The rsync stuff is because MacOS performance is terrible for docker volumes. I would love to drop it, but last I looked we just couldn't.

The performance got better over time, and with a few tweaks we can drastically reduce the I/O required anyhow. If we can actually leverage go's cache with modules enabled we can also cache the modules within a docker volume instead of bind mounting vendor and aggressively cache most of the build.

[This](https://github.com/kubernetes-sigs/kind/blob/e8513836ae23b59c660e290043380e193f10d07e/Makefile#L71) runs regularly on a mac building a go binary w/ caching and achieves roughly native performance on mac, incremental compiles are very quick.

> I don't disagree that bash may be at the end of it's runway, but I question python as an alternative.

FWIW the Google shell style guide does [suggest using python](https://google.github.io/styleguide/shell.xml?showone=When_to_use_Shell#When_to_use_Shell) when you shouldn't use bash... And we hit every point on when it suggests to switch :upside_down_face: 

Go is another possibility in the OP.

**Comment by jberkus** [comment]
Just a note that Python is not installed on Windows by default.  So this would make installing python a requirement for Kubernetes Windows users.

/cc @kubernetes/sig-windows-misc

**Comment by cblecker** [comment]
Contributors, not users. This is focused on the development and build systems. We already have some python in the repo, so a `make verify` will fail in any environment that lacks python.

**Comment by stevekuznetsov** [comment]
POSIX is a fairly onerous and difficult standard to adhere to. Why?

**Comment by BenTheElder** [comment]
Also note that _none_ of our builds work on default installs. They require some mixture of docker, bash, make, go, bazel, etc. I don't think Docker is installed by default on most operating systems...

I also don't think there's been any real effort to _build_ from Windows, but we build _to_ Windows from Linux in CI. This shouldn't affect users.

**Comment by BenTheElder** [comment]
> POSIX is a fairly onerous and difficult standard to adhere to. Why?

It's not particularly onerous for _small_ scripts. Anything not small probably shouldn't be in shell.

**Comment by stevekuznetsov** [comment]
> It's not particularly onerous for small scripts. Anything not small probably shouldn't be in shell.

Right by why disallow `[[` and all the niceties of it and force everyone to use the POSIX-compliant `[` or `test`? The latter have gotchas that make using them tricky.

**Comment by BenTheElder** [comment]
@stevekuznetsov 🤷‍♂ hasn't been a huge problem so far, the _small_ script ideally should just be non-branchy aliases around the bigger tools anyhow.

---

Another datapoint: nearly all of the tooling doesn't work on a stock macOS install anyhow (even ignoring docker).
- some of it requires more recent bash
- some of it requires md5sum
- some of it requires gnu-sed
- ... I'm sure we're installing other utilities

The current scripts have pretty large dependency surface on the host anyhow.

**Comment by justaugustus** [comment]
/area release-eng
/help
/milestone v1.16
/priority important-longterm
cc: @kubernetes/bash-firefighters

**Comment by cblecker** [comment]
/remove-help I don't think we're quite yet at a consensus point, and this isn't just a release thing.

**Comment by cblecker** [comment]
/remove-help
I don't think we're quite yet at a consensus point, and this isn't just a release thing.

**Comment by makoscafee** [comment]
@cblecker Hello! I'm bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by BenTheElder** [comment]
No, there's not consensus yet. Can we get milestones removed from this
entirely?

**Comment by zouyee** [comment]
Due to bash compatibility issues, the new version features cannot be supported in the old version. Do I need to add a version check for bash and the minimum version requirements?
issue:
 https://github.com/kubernetes/kubernetes/issues/85277
 https://github.com/kubernetes/kubernetes/issues/82605
 https://github.com/kubernetes/kubernetes/pull/85631
xref:
 https://stackoverflow.com/questions/6047648/bash-4-associative-arrays-error-declare-a-invalid-option/43948526
https://github.com/confluentinc/confluent-cli/issues/15
https://unix.stackexchange.com/questions/498280/target-system-bash-give-me-parameter-error-when-i-use-local-n-how-am-i-allow

**Comment by BenTheElder** [comment]
There is a minimum version check in some scripts, we already have a utility for it.

That won't avoid compatibility issues really though. Careful scripting can, as can not using bash for large programs :-)
